### PR TITLE
236 add how to use the mirror pipeline section to onboarding doc

### LINKED ISSUE #236: Add "How to use the MIRROR Pipeline" section to onboarding doc
Acceptance criteria:
- A section is added to `mirror onboarding.md` explaining:
   - How to do a training run using existing classes
   - How to use other models/datasets/etc than the ones that are currently in the repo (i.e. explain that implementing the interfaces is the way to do it)

### PR DESCRIPTION
**Pull request by JansenNye**
Tested using markdown file preview I guess
Closes #236

### CODE REVIEW

File: docs/mirror onboarding.md [hunk 1]
```diff
@@ -132,37 +132,97 @@ Vim is the default editor for commit message files (e.g. git merge).
...
+```yaml
+preprocessor:
+    class_path: <class_path> # e.g. MirrorLlamaPreprocessor
+```
+
+To preprocess "lazily"/on-the-fly (slower, saves memory):
+
+```yaml
+do_preprocess: True # False by default
+```
```
Comment (morrisoncheney) [inline docs/mirror onboarding.md]: This is actually not correct, you need `do_preprocess: False` in order to preprocess on the fly.

File: docs/mirror onboarding.md [hunk 2]
```diff
@@ -132,37 +132,97 @@ Vim is the default editor for commit message files (e.g. git merge).
...
+```yaml
+data: # Training set 
+  class_path: WikitextDataset
+  init_args:
+    split: train # Use the dataset's `train` portion
+    skip: <i> # Optionally, skip the first <i> examples from the selected split 
+    head: <j> # Use the first <j> examples (after skipping)
+    # OR
+    # start_fraction: 0.0
+    # end_fraction: 0.8 # Use the first 80% of the selected dataset/split
```
Comment (morrisoncheney) [inline docs/mirror onboarding.md:282]: I think the start fraction stuff is only implemented on branch 86 if I'm not mistaken.

File: docs/mirror onboarding.md [hunk 3]
```diff
@@ -132,37 +132,97 @@ Vim is the default editor for commit message files (e.g. git merge).
...
+    split: test # Use the dataset's `test` portion
+    ... # etc
+```
+
+Next, pick your SLURM job/training run settings. 
+
+```yaml
+slurm:
+  job_type: "compute" # "local" "local-download"
+  time: "01:00:00" # Timeout limit
```
Comment (morrisoncheney) [inline docs/mirror onboarding.md]: optional: it could be helpful to mention that the requeue callback will continue the job even if the time limit is hit by creating another slurm job submission. That may not need to go right here though.

**Review by morrisoncheney — CHANGES_REQUESTED** [review]
There were a couple of little things to review.

File: docs/mirror onboarding.md (hunk 1 above)
Comment (JansenNye) [reply docs/mirror onboarding.md]: Whoops

File: docs/mirror onboarding.md (hunk 2 above)
Comment (JansenNye) [reply docs/mirror onboarding.md:282]: Are we not going to be merging that branch? If not, we should probably make that same change on a ticket that will be merged, since that's a really useful feature for text datasets without specific splits

File: docs/mirror onboarding.md (hunk 3 above)
Comment (JansenNye) [reply docs/mirror onboarding.md]: Good idea

File: docs/mirror onboarding.md (hunk 2 above)
Comment (morrisoncheney) [reply docs/mirror onboarding.md:282]: I thought Koby had said we weren't gonna, but I could be wrong. 
In any case I think it would make sense to add these to this doc once they are merged to main, rather than having them in here while they "don't exist" for users to access.

I agree. That definitely is a useful feature and keeps users from having to do the extra work involved in figuring out how many rows are in their dataset and such, definitely seems worth implementing in the pipeline.

**Review by morrisoncheney — CHANGES_REQUESTED** [review]
Just the one comment that I have on there.

**Review by morrisoncheney — APPROVED** [review]
Lookin good!

File: docs/mirror onboarding.md [hunk 4]
```diff
@@ -132,37 +132,98 @@ Vim is the default editor for commit message files (e.g. git merge).
     - Useful for preparing data separately before running a training job
     - Requires `--data` and `--preprocessor` to be specified (either in the config file or as command-line arguments)
 
+## How to use the MIRROR Pipeline
+
+It's highly recommended to use YAML config files rather than passing in arguments through the command line. Thus, this section will outline how to use the MIRROR Pipeline through the lens of creating this config file.
```
Comment (Plyb) [inline docs/mirror onboarding.md]: This feels like it's missing context. I'd maybe start with something like just telling them the command to run the pipeline using one of the demo configs. Then you can jump in to telling them about the different parts of the config files and how to edit them. The goal here is for someone to be able to just read the first part of this section and make sure they've got the pipeline set up correctly.

File: docs/mirror onboarding.md [hunk 5]
```diff
@@ -132,37 +132,98 @@ Vim is the default editor for commit message files (e.g. git merge).
...
+        initialization:
+            # Pretrained weights:
+            <model_spec> # 3.2-1B | 3.2-1B-Instruct
+            # OR custom config:
+            vocab_size: <vocab_size>
+            hidden_size: <hidden_size> # etc
+
+        # If using GPT2:
+        weights: <weights> # pretrained | random
+    # Etc/as appropriate for other models
```
Comment (Plyb) [inline docs/mirror onboarding.md]: As someone who has never seen this before, how do I know what is "appropriate" for different models? We should tell them what needs to go into these YAML configs in general (i.e. the constructor parameters and how those are translated by jsonargparse)

File: docs/mirror onboarding.md [hunk 6]
```diff
@@ -132,37 +132,98 @@ Vim is the default editor for commit message files (e.g. git merge).
...
+```yaml
+preprocessor:
+    class_path: <class_path> # e.g. MirrorLlamaPreprocessor
+```
+
+To preprocess up front rather than on-the-fly:
+
+```yaml
+do_preprocess: True # False by default
+```
```
Comment (Plyb) [inline docs/mirror onboarding.md]: Could maybe move this down? As someone just starting, I think this would feel like noise. I'd probably be wondering about datasets at this point

File: docs/mirror onboarding.md (hunk 2 above)
Comment (Plyb) [reply docs/mirror onboarding.md:282]: yeah, seems like a useful feature. I was planning on merging them in. No real preference on keeping this as is for now or removing it and re-adding it when the feature is merged.

**Review by Plyb — CHANGES_REQUESTED** [review]
One suggestion for how to structure this: think about what questions a person might have as their getting set up and answer those in order. For example, the first question is probably just "did I install this correctly", so letting them know the easiest way to just do a training run with a demo file. Maybe sometime later I'm wondering how to do validation and test datasets, so have a section titled "How to use Val and Test Sets" or something like that. Frame it all around answering "How do I do X"

In all cases though, keep in mind that someone reading it has no context. They don't know anything about our pipeline.

File: docs/mirror onboarding.md [hunk 7]
```diff
@@ -220,6 +220,127 @@ Vim is the default editor for commit message files (e.g. git merge).
...
+Note that you'll need to [use an h200 GPU](#setting-up-slurm-jobtraining-run-settings) in order to have enough VRAM to run Llama with pretrained weights, so we usually use a smaller custom config of Llama, like the one defined in [demo_fit_config.yaml](../configs/demo_fit_config.yaml).
+
+To use GPT-2:
+
+```yaml
+model:
+    class_path: MirrorGPTModel 
+    init_args: 
+        weights: <weights> # pretrained | random
+```
```
Comment (Plyb) [inline docs/mirror onboarding.md:267]: Does this accidentally give the impression that there are only two models to use?

File: docs/mirror onboarding.md [hunk 8]
```diff
@@ -220,6 +220,127 @@ Vim is the default editor for commit message files (e.g. git merge).
...
+        initialization:
+            # Pretrained weights:
+            <model_spec> # 3.2-1B | 3.2-1B-Instruct
+
+            # OR custom config:
+            vocab_size: <vocab_size>
+            hidden_size: <hidden_size> # etc
+```
+
+Note that you'll need to [use an h200 GPU](#setting-up-slurm-jobtraining-run-settings) in order to have enough VRAM to run Llama with pretrained weights, so we usually use a smaller custom config of Llama, like the one defined in [demo_fit_config.yaml](../configs/demo_fit_config.yaml).
```
Comment (Plyb) [inline docs/mirror onboarding.md]: I think h200, h100, a100, or b200s should work. Might want to make this more generic

File: docs/mirror onboarding.md [hunk 9]
```diff
@@ -220,6 +220,127 @@ Vim is the default editor for commit message files (e.g. git merge).
...
+
+You should see the output `Submitted batch job <number>`. If this is your first time submitting a training run, you may be prompted to [log in to huggingface](#initial-access-setup) first to download required resources like the Llama model weights.
+
+That command uses a demo config YAML file to specify the settings for the training run. To customize a training run, you can either use a config file or pass in each argument and its value through the command line, e.g.:
+
+```
+python src/main.py fit --data.class_path WikitextDataset --data.head 10 --model.class_path MirrorLlamaModel
+```
+
+It's recommended to use config files rather than passing in arguments through the command line. Thus, this section will outline how to use the MIRROR Pipeline by walking you through creating this config file, section by section.
```
Comment (Plyb) [inline docs/mirror onboarding.md]: I still think somewhere it would be worth mentioning how the yaml files relate to the constructor arguments. I want people to know that if they want to know how to initialize a model/dataset/etc, they can just go look at the class and see what it takes in.

File: docs/mirror onboarding.md [hunk 10]
```diff
@@ -220,6 +220,127 @@ Vim is the default editor for commit message files (e.g. git merge).
...
+```yaml
+val_data: # Validation set
+  class_path: WikitextDataset
+  init_args:
+    split: validation # Use the dataset's `validation` portion
+    ... # etc
+
+val_check_interval: <k> # Perform validation every <k> epochs
+
+test_data: # Test set
```
Comment (Plyb) [inline docs/mirror onboarding.md]: Might be worth mentioning that the test set will be run at the end of the training run

File: docs/mirror onboarding.md [hunk 11]
```diff
@@ -220,6 +220,127 @@ Vim is the default editor for commit message files (e.g. git merge).
...
+  mem_per_cpu: "128G" # Memory allocated per CPU core
+  output: "slurm_logs/%j.out" # Path for job stdout log (%j = job ID)
+  open_mode: "append" # Append to log file instead of overwriting
+  signal: "SIGHUP@90" # Send SIGHUP signal 90 seconds before timeout to allow graceful shutdown
+  requeue: true # Automatically requeue job if it is preempted
+  qos: <cluster> # Specific GPU cluster/config, e.g. dw87, test
+
+epochs: 1 # Num of passes through the training dataset
+batch_size: 1 # Num of samples in each batch 
+device: cuda # `cuda` for GPU, `cpu` for CPU
```
Comment (Plyb) [inline docs/mirror onboarding.md:340]: do we actually have a `device` parameter?

File: docs/mirror onboarding.md [hunk 12]
```diff
@@ -220,6 +220,127 @@ Vim is the default editor for commit message files (e.g. git merge).
...
+
+```yaml
+do_preprocess: True # False by default
+``` 
+
+#### Setting up SLURM job/training run settings
+
+```yaml
+slurm:
+  job_type: "compute" # "local" "local-download"
```
Comment (Plyb) [inline docs/mirror onboarding.md]: maybe worth either describing what these options are here, or linking to somewhere that does

File: docs/mirror onboarding.md [hunk 13]
```diff
@@ -220,6 +220,127 @@ Vim is the default editor for commit message files (e.g. git merge).
...
+  time: "01:00:00" # Timeout limit - note that the requeue callback will continue 
+  # the job if the time limit is reached by submitting another SLURM job
+  gpus_per_node: <gpu_type>:<num_gpus> # GPUs, low -> high capability: p100 | a100 (dw87 cluster) | h200
```
Comment (Plyb) [inline docs/mirror onboarding.md]: The gpu list here isn't exaustive
```suggestion
  gpus_per_node: <gpu_type>:<num_gpus> # GPUs, low -> high capability: p100 | a100 (dw87 cluster) | h200 | etc
```

**Review by Plyb — CHANGES_REQUESTED** [review]

File: docs/mirror onboarding.md (hunk 11 above)
Comment (JansenNye) [reply docs/mirror onboarding.md:340]: Yes, we had had that whole issue previously where we were setting `device: cpu` due to the false CUDA unavailable warning

Comment (Plyb) [reply docs/mirror onboarding.md:340]: ah, I see it now. I got confused because it wasn't an argument of the `fit` subcommand function

**Review by Plyb — APPROVED** [review]
